### PR TITLE
Download ais

### PR DESCRIPTION
**Pull request by MorganMasters**

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -38,6 +38,7 @@ install_requires =
     onc
     sqlalchemy
     spans
+    wget
```
Comment (Jacob-Stevens-Haas) [inline setup.cfg]: Let's use `requests`, not `wget`.  `wget` seems to be dead content as it's last update was 2015 and it's [bitbucket link is daed](https://pypi.org/project/wget/)

tl;dr: replace `wget.download` with `requests.get`

File: tehom/downloads.py [hunk 2]
```diff
@@ -32,6 +34,9 @@
 
 from . import _persistence
 
+# from zipfile import ZipFile
```
Comment (Jacob-Stevens-Haas) [inline tehom/downloads.py]: delete this and extra lines

File: tehom/downloads.py [hunk 3]
```diff
@@ -80,8 +87,72 @@ def _download_ais_to_temp(year: int, month: int, zone: int) -> Path:
...
+            "https://coast.noaa.gov/htdata/CMSP/AISDataHandler/{year}/"
+            f"{month}_{months[month]}_{year}/Zone{zone}_{year}_{month}.zip"
+        )
+
+    if isinstance(zone, int) and zone < 10:
+        zone = f"0{zone}"
+    filepath = _persistence.AIS_TEMP_DIR / f"{year}_{month}_{zone}.zip"
+    if not filepath.exists():
+        try:
+            print(f"Downloading data for {year} {month}, Zone {zone}...")
```
Comment (Jacob-Stevens-Haas) [inline tehom/downloads.py]: I'm not a fan of leaving in diagnostic print statements.  If they're potentially worth including, add 
`logger = logging.getLogger(__name__)` at the beginning of the module, after import statements.  Then, instead of `print(...)`, use `logger.info(...)`.

This allows users (or applications using this library) to suppress or redirect outputs to a log.

File: tehom/downloads.py [hunk 4]
```diff
@@ -80,8 +87,72 @@ def _download_ais_to_temp(year: int, month: int, zone: int) -> Path:
...
+            wget.download(url, filepath)
+            with open("downloadRecord.txt", "wb") as f:
+                f.write(f"{year}, {month}, {zone}, success")
+        except urllib.error.HTTPError:
+            with open("downloadRecord.txt", "wb") as f:
+                f.write(f"{year}, {month}, {zone}, failure")
```
Comment (Jacob-Stevens-Haas) [inline tehom/downloads.py]: Let the error percolate up.  From a theory perspective, this function doesn't know how it's being used.  If this function experiences an HTTPError, it probably can't do it's main job and the **caller** of this function should determine how to handle the error.  In other words, ZOP 10-12.

Logging would be an ok addition, but I don' think it's necessary here.  So just drop the whole exception catching block.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Hey Morgan, congrats on first pull request! :partying_face: :champagne:  Thanks for writing through this.  A few small changes noted below, but also one big one: you need to also write a test to tehom/tests/test_downloads.py.  It would look something like:
```python
@pytest.mark("slow")
def test_download_ais_to_temp(declare_stateful):
    <download a year-month-zone combo that has as few records as possible, e.g. far out to the ocean>
    path = Path(<path to target file.zip>)
    assert path.exists()
```

As for other general guidance, I'd recommend reading [Google's PR guide](https://google.github.io/eng-practices/review/) and [what makes a good commit message](https://chris.beams.io/posts/git-commit/).

File: tehom/downloads.py (hunk 2 above)
Comment (MorganMasters) [reply tehom/downloads.py]: This is in for unzip-ais, which I think I have since pushed
 
If not, that'll happen shortly

File: setup.cfg (hunk 1 above)
Comment (MorganMasters) [reply setup.cfg]: wget is a far more basic command, simply a wrapper on the Linux wget command, plus easier usage in Python.

Are you sure you want to make this change?

Comment (Jacob-Stevens-Haas) [reply setup.cfg]: yes.  `requests` is the standard here.  Need Windows and Mac compliance.  Python wget wrapper doesn't have any documentation, so I don't know how to pass any of the dozen `wget` command line arguments via python.

File: tehom/tests/test_downloads.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from tehom import _downloads, _persistence
+
+
+def test_download_ais_to_temp(declare_stateful):
```
Comment (Jacob-Stevens-Haas) [inline tehom/tests/test_downloads.py:7]: Decorate this with `@pytest.mark("slow")`

File: tehom/tests/test_downloads.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    year = 2014
+    month = 1
+    zone = 1
+    _downloads._download_ais_to_temp(year, month, zone)
+    path = _persistence.AIS_TEMP_DIR / f"{year}_{month}_{zone}.zip"
```
Comment (Jacob-Stevens-Haas) [inline tehom/tests/test_downloads.py]: since `_download_ais_to_temp` returns a path, you can use that value as `path`.

File: tehom/tests/test_downloads.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from tehom import _downloads, _persistence
```
Comment (Jacob-Stevens-Haas) [inline tehom/tests/test_downloads.py]: This line is failing the build: 
```
ImportError while importing test module '/home/runner/work/tehom/tehom/tehom/tests/test_downloads.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.8.11/x64/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tehom/tests/test_downloads.py:1: in <module>
    from tehom import _downloads, _persistence
E   ImportError: cannot import name '_downloads' from 'tehom' (/home/runner/work/tehom/tehom/tehom/__init__.py)
```
tl;dr: it's `downloads`, not `_downloads`

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Hey, thanks Morgan.  Should be a few quick fixes.  Don't forget to replace wget with requests.

**Comment by Jacob-Stevens-Haas** [comment]
All good - just mark the test with `@pytest.mark("slow")` and we can merge.

File: tehom/tests/test_downloads.py (hunk 5 above)
Comment (MorganMasters) [reply tehom/tests/test_downloads.py:7]: build failed:

tehom/tests/test_downloads.py:6: in <module>
    @pytest.mark("slow")
E   TypeError: 'MarkGenerator' object is not callable

It was for this reason that I omitted the decorator on my push. The docs say something about defining the 'slow' marker in a pytest.ini file: https://docs.pytest.org/en/6.2.x/mark.html

Comment (Jacob-Stevens-Haas) [reply tehom/tests/test_downloads.py:7]: Weird, [it is defined as a marker](https://github.com/Jacob-Stevens-Haas/tehom/blob/960538e607e8134992425e15bd12ed4cbe087355/pyproject.toml#L37).  Let me take a look at this.

Comment (Jacob-Stevens-Haas) [reply tehom/tests/test_downloads.py:7]: Yeah, if you run `pytest --markers`, you can see that it should be written:
```python
@pytest.mark.slow
def test_...
```
not
```python
@pytest.mark("slow")
def test_...
```
My mistake.
